File: 0872-leaf-similar-trees/0872-leaf-similar-trees.java
class Solution {
    List<Integer> myList=new ArrayList();
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        inorderStore(root1);
        for(int n:myList) System.out.print(n+" ");
        boolean res=inorderCheck(root2);
        boolean s=myList.isEmpty();
        return res && s;
    }
    void inorderStore(TreeNode root){
        if(root==null) return;
        if(root.left==null && root.right==null){
            myList.add(root.val);
            return;
        }

        inorderStore(root.left);
        inorderStore(root.right);
    }
    boolean inorderCheck(TreeNode root){
        if(root==null) return true;
        if(root.left==null && root.right==null){
            if(!myList.isEmpty() && myList.get(0)==root.val){
                myList.remove(0);
                return true;
            }
            else return false;
        }
        boolean st1=inorderCheck(root.left);
        boolean st2=inorderCheck(root.right);
        return st1 && st2;
    }
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        string s1, s2;
        dfs(root1, s1);
        dfs(root2, s2);
        return s1 == s2;
    }
    
    void dfs(TreeNode* root, string& s) {
        if(root == NULL) return;
        if(root->left == NULL and root->right == NULL) 
            s += to_string(root->val) + "-";
        dfs(root->left, s);
        dfs(root->right, s);
    }
};